parlai/tasks/ccpe/agents.py

```diff
@@ -3,150 +3,106 @@
-                        entry.append(False)
-                        currAssistantEp.append(entry)
-                        asssistantCnt += 1
-
-            currUserEp[-1][5] = True
-            currAssistantEp[-1][5] = True
-            self.userData.append(currUserEp)
-            self.assistantData.append(currAssistantEp)
-
-        self.data = self.assistantData + self.userData
-
-    def get(self, episode_idx, entry_idx=0):
-        ep = self.data[episode_idx]
-        entry = ep[entry_idx]
-        action = {
-            'id': entry[0],
-            'text': entry[1],
-            'labels': entry[2],
-            'textSegments': entry[3],
-            'labelSegments': entry[4],
-            'episode_done': entry[5],
-        }
-        return action
-
-    def share(self):
-        shared = super().share()
-        shared['data'] = self.data
-        return shared
-
-
-class CCPEAssistantTeacher(CCPEAllTeacher):
-    def _setup_data(self):
-        super()._setup_data()
-        self.data = self.assistantData
-
-
-class DefaultTeacher(CCPEAllTeacher):
+        # do a 80:10:10 train/valid/test split
+        full_episode_count = len(json_data) // 10
+        if fold == 'train':
+            json_data = json_data[: full_episode_count * 8]
+        elif fold == 'valid':
+            json_data = json_data[full_episode_count * 8 : full_episode_count * 9]
+        elif fold == 'test':
+            json_data = json_data[full_episode_count * 9 :]
+
+        episodes = []
+        for conversation in json_data:
+            episode = []  # Assume all turns are alternating with ASSISTANT first
+            idx = 0
+            utts = conversation["utterances"]
+            if utts[0]["speaker"] != "ASSISTANT":
+                # Make a dummy assistant turn and add a user turn
+                episode.append({"speaker": "ASSISTANT", "text": "", "segments": []})
+                turn, idx = self._get_turns_of_speaker(utts, idx, "USER")
+                episode.append(turn)
+            while idx < len(utts):
+                turn, idx = self._get_turns_of_speaker(utts, idx, "ASSISTANT")
+                episode.append(turn)
+                turn, idx = self._get_turns_of_speaker(utts, idx, "USER")
+                episode.append(turn)
+            if len(episode) % 2 != 0:
+                # Add a dummy user turn so we don't need to worry about labels
+                episode.append({"speaker": "USER", "text": "", "segments": []})
+            episodes.append(episode)
+        return episodes
+
+
+class CcpeAssistantTeacher(_Abstract):
+    def setup_data(self, fold):
+        episodes = self._load_data(fold)
+        for episode in episodes:
+            first = True
+            for i in range(len(episode) // 2):
+                assistant_turn = episode[2 * i]
+                user_turn = episode[2 * i + 1]
+                if len(assistant_turn["text"]) == 0:
+                    continue  # trains fail otherwise
```

Comment:
so you're adding in a dummy example above, but then you're skipping it in these teachers?

Reply:
I'm adding a dummy turn above so that the format returned from `_load_data(fold)` has the same expectations of 

> first turn = ASSISTANT
> second turn = USER
> ... and so on

since some of the conversations in the original dataset start with the USER. This makes it easier to specify which turns of the episode should be used by CcpeAssistantTeacher and which should be used by CcpeUserTeacher

(I can see now how that's not super obvious at first glance... can add more comments.)

Reply:
I need to skip it here because a part of `self_observe()` errors out if an empty text string was passed in.

Reply:
yeah that can be solved by putting e.g. the `__SILENCE__` token